packages/app/src/App.tsx

```diff
@@ -98,12 +99,19 @@ const app = createApp({
   },
   components: {
     SignInPage: props => {
+      const analytics = useAnalytics();
```

Comment:
Before going further, want to get your opinion on this @freben or @Rugvip or @jhaals.  

Goal is to allow folks to apply whatever pseudonymization process they want/need (if any) to the user ID being set.  ...This seemed like the simplest approach to doing this.  ...But it means there won't be any automatic instrumentation of user-level info.

Whaddya think?

Reply:
Adding `setUser` has pretty big impact and isn't really interesting to consumers, think we could squeeze it into the instantiation of the concrete `AnalyticsApi` instead?

```ts
  createApiFactory({
    api: analyticsApiRef,
    deps: { configApi: configApiRef, identityApi: identityApiRef },
    factory: ({ configApi, identityApi }) =>
      GoogleAnalytics.fromConfig(configApi, {
        userIdentifierPromise: identityApi.getBackstageIdentity()
          .then(identity => identity.userEntityRef /* insert custom logic here */)
      }),
  }),
```

Reply:
Reasonable!  But when I was playing around with this on an install of Backstage that's not this monorepo, I found this approach led to `userIdentifierPromise` just rejecting with `Tried to access IdentityApi before app was loaded` and no opportunity to retry (without coding the retry right there).

Reply:
> Setters in the APIs themselves aren't unheard of, but I feel it's a bit extra bad when you're expecting to only set it once at app boot, and never touch it again.

Agree. But it seems important enough to explicitly make part of the API surface area in some way and we don't really have a nice, standard way of doing that with constructor args for APIs.

Maybe we decouple the tracker returned by `useAnalytics()` and the `AnalyticsApi` definition so we don't expose the setter to plugin code? And maybe we add something like a fluent interface to the API?

e.g.

```ts
SomeAnalyticsApi.fromConfig(configApi).withUserId(someIdPromise);
```

Still feels weird to take a promise to a string as the argument in any case, but I guess it's not really avoidable.

> A slight spin that might be a bit safer would be to add a `waitForSignIn` option or something like that to `getBackstageIdentity`, that way we avoid the above risks and it's simple to implement for all APIs in the `AppIdentityProxy`.

I'll take a look at this.

Reply:
I'd say our standard way of doing this right now is either extra options to `.create`, or a second parameter with extra options to `.fromConfig`. Tbh I find that fluid interfaces in JS/TS are really only useful when you're building queries, adding middlewares, or that kinda thing. Basically more complex options where the chained method is typically overloaded. Most of the time an options object is just as powerful and easy to use, formats better, is easier to implement and test, and avoids confusion like whether the original object is immutable or is mutated.

👍 trying out the `waitForSignIn` route

Reply:
> I'd say our standard way of doing this right now is either extra options to .create, or a second parameter with extra options to .fromConfig

We just don't usually (ever?) enshrine those methods or their extra arguments in the API interface itself; I'm just trying to find the least awkward way to incorporate user ID into the interface.

Reply:
Goal is to standardize how `AnalyticsApi` implementors receive a user ID and for app integrators to provide such an ID:

- If a plugin developer wants to come in and create an `AmplitudeAnalyticsApi` implementation, there should be a clear, prescribed way to accept a user ID.
- So that: if an app integrator wants to migrate from `GoogleAnalyticsApi` to `AmplitudeAnalyticsApi`, there's little, if any, changes necessary, beyond providing the alternate API implementation.

It feels really _wild west_ for it to basically only be by convention, and for the convention to be "accept a promise to a string in the constructor or static factory method"

Reply:
Aaaalright, I see where you're coming from.

After playing with adding `waitForSignIn` to the API...  It felt really weird to introduce something in the interface, even if optional, that really only one concrete implementation would ever care about.  So I tried the first approach.

Gonna work on that one isolated change here, first: https://github.com/backstage/backstage/pull/8945

Reply:
(301 redirect: https://github.com/backstage/backstage/pull/8811#:~:text=which%20relies%20on-,%238945,-iamEAP%20closed%20this)